### DISCUSSION TITLE
Remember git branch in updates

### DISCUSSION THREAD

**Discussion by ruestefa** [copier-org/copier#901] [Q&A]
### Goal

I am experimenting with how to best handle templates that differ only in a few aspects, e.g., a for a pure-Python project and one with `f2py` support. A single git repo with multiple main branches, one for each flavor of the template (e.g., `main`, `main-f2py`), seem like a good way to handle this.

### Problem

In principle, this works well by specifying the branch with `--vcs-ref`. Unfortunately, there does not seem to be a way for copier to remember the branch during `copier update`s, i.e., users always need to pass the correct `--vcs-ref` flag, which is not feasible.

_(Right now, the only way I see to manage multiple related templates in one repo while not burdening users with `--vcs-ref` during updates is to create a fork for each flavor of the template, which seems unwieldy, especially if the template is hosted on multiple platforms.)_

### Possible solutions

- Add an option to remember `--vcs-ref` in `.copier-answers.yml` and, if present, reuse that value during updates.
- Add support for branches to the URL syntax, e.g., with a `@<branch>` suffix as suggested in #227.

### _What else I've tried_

_Copier support multiple templates for a project, e.g., a pure-Python base template with a partial template with `f2py` modifications. This works but requires exposing the user to multiple templates. To avoid that, I've experimented with "meta templates", i.e., templates with two `tmpl/` layers, that can be combined into complete templates. This does work, but complicates maintenance and feels more like a hacky workaround than a proper solution._

**Comment by yajo** [comment]
Hello and welcome!

Copier is indeed a bit opinionated about how are you supposed to develop
your templates. The canonical way is: a main branch, PEP-440 tags for
releases, and branches are useful for testing volatile changes until merged
in the main branch.

Why is that so? Because of updates. For them to work, we need to be able to
reproduce last render, new one, and compare both with current state of the
project. For future-proof updates, you need migrations. For migrations, you
need tags. And tags are global for git, not per branch.

So, if we use moving refs (like branches), we lose all of that. To give
your desired use case appropriate support, we'd need to cut down so many
features that it would be more practical for you to use other tools that
don't care so much about uplates and so don't worry about these things.

But we want you here, so what could you do to solve your need?

Option 1: use 2 repos instead of 2 branches. If you fork one from the
other, you'll be able to share commits back and forth with not much more
effort.

Option 2 (more canonical): single repo, single branch. You just add a
question to your questionary:

```yaml
# copier.yaml
f2py:
  default: false
```

Then use that variable in your templates to add support for f2py
conditionally.

I think option 2 is better because it will remember the answer for updates
and you'll have the full potential of copier because you will follow its
mental model. But it's up to you.

As you can see, it's almost impossible to support your proposal, so I'll
close this one. Let me convert it to discussion, as this will be more
helpful there.

Thanks!

**Reply by ruestefa** [reply]
Hi, thank you for your quick and detailed response! I see your point, so it indeed probably doesn't make sense to go down my proposed road.

I've meanwhile found a workable solution in creating a separate empty repo (not a fork) per template flavor and only push the respective branch there. This approach only requires one additional push per branch to propagate changes while all the work can still be done in a centralized repo.

As for your one-repo-one-branch proposal with conditional content: We used conditionals for some things in an earlier version of the template, but this tends to litter the files (and file names) with jinja2 directives. I find keeping track of the differences in git branches with standard git tools much simpler and more transparent.

> it would be more practical for you to use other tools that don't care so much about uplates...

Updates were the single reason I switched from cookiecutter to copier in the first place (and they work well), so I won't give up on that. :-)

Thanks for your work and prompt support!
